test/run_test.py

```diff
@@ -934,76 +940,190 @@ def run_test_retries(
+            return 1, False
 
         env = try_set_cpp_stack_traces(env, command, set=False)
-        if ret_code != 0:
-            num_failures[current_failure] += 1
 
-        if ret_code == 0:
-            # Rerunning the previously failing test succeeded, so now we can
-            # skip it and move on
-            sc_command = f"--scs={stepcurrent_key}"
-            print_to_file(
-                "Test succeeeded in new process, continuing with the rest of the tests"
+        # Remove prev_run from to_run
+        cache["to_run"] = [
+            [x, y]
+            for x, y in cache["to_run"]
+            if x not in [y for y, _ in cache["prev_run"]]
+        ]
+
+        if ret_code != 0 and cache["pytest_previous_status"] == 0:
+            # Failed after exiting
+            # Remove old xml since it is incorrect
+            os.remove(test_report_path)
+            if len(cache["prev_run"]) == 1:
+                make_manual_xml(
+                    invoking_file,
+                    cache["ended_at"],
+                    duration,
+                    get_last_lines() + "\nFailed at exit?",
+                )
+
+            # Rerun all tests on their own to figure out which one
+            # caused the process to fail at exit
+            cache["to_run"] = [[x, get_rerun(y)] for x, y in cache["prev_run"]] + cache[
+                "to_run"
+            ]
+
+        elif cache["pytest_previous_status"] == "no xml":
+            # Segfault or something similar
+            # Manually create xml entry
+            make_manual_xml(
+                invoking_file, cache["ended_at"], duration, get_last_lines()
+            )
+
+            # Rerun everything prior to the segfaulting test to generate xml
+            i, t = next(
+                (i, t)
+                for i, t in enumerate(cache["prev_run"])
+                if t[0] == cache["ended_at"]
+            )
+            cache["to_run"] = (
+                cache["prev_run"][:i]
+                + [[cache["ended_at"], get_rerun(t[1], failed=True)]]
+                + cache["prev_run"][i + 1 :]
+                + cache["to_run"]
             )
-        elif num_failures[current_failure] >= 3:
+
+        elif cache["pytest_previous_status"] != 0:
+            # Failed last run, so we need to rerun the last test
+            i, t = next(
+                (i, t)
+                for i, t in enumerate(cache["prev_run"])
+                if t[0] == cache["ended_at"]
+            )
+            cache["already_ran"] = cache["already_ran"] + cache["prev_run"][:i]
+            cache["to_run"] = (
+                [[cache["ended_at"], get_rerun(t[1], failed=True)]]
+                + cache["prev_run"][i + 1 :]
+                + cache["to_run"]
+            )
+
+        elif cache["pytest_previous_status"] == 0:
+            # Success
+            cache["already_ran"].extend(cache["prev_run"])
+
+        if len(cache["to_run"]) > 0 and cache["to_run"][0][1] == "failed":
```

Comment:
nit: named tuples could let this become a more readable:

```suggestion
        if len(cache["to_run"]) > 0 and cache["to_run"][0].status == "failed":
```